docs/CompilerPerformance.md

```diff
@@ -0,0 +1,1142 @@
+machine-readable and aggregates all the jobs in a multi-job compilation, and
+there's a post-processing script `utils/process-stats-dir.py` to work with these
+files in aggregate.
+
+#### Post-processing tools for diagnostics
+
+If you dump diagnostic output using `-stats-output-dir <dir>`, the resulting
+files in `<dir>` will be simple JSON files that can be processed with any
+JSON-reading program or library, such as `jq`. Alternatively, a bulk-analysis
+script also exists in `utils/process-stats-dir.py`, which permits a variety of
+aggregation and analysis tasks. The most useful tasks revolve around setting and
+comparing against so-called "baselines": merge-friendly CSV files that summarize
+all the counters from all the stats files in a stats-dir.
+
+Here is an example of how to use `-stats-output-dir` together with
+`utils/process-stats-dir.py` to analyze the difference in compilation
+performance between two compilers, say `${OLD}/swiftc` and `${NEW}/swiftc`:
+
+```
+$ mkdir stats
+$ ${OLD}/swiftc -stats-output-dir stats test.swift
+$ utils/process-stats-dir.py -set-csv-baseline baseline.csv stats
+making new baseline baseline.csv
+
+$ rm stats/*
+$ ${NEW}/swiftc -stats-output-dir stats test.swift
+$ utils/process-stats-dir.py -compare-to-csv-baseline baseline.csv stats
+old     new     delta_pct       name
+1402939 1430732 1.98    AST.NumASTBytesAllocated
+7       0       -100.0  AST.NumUsedConformances
+232     231     -0.43   Constraint solver largest system.LargestNumStatesExplored
+42      41      -2.38   Constraint solver largest system.LargestNumTypeVariableBindings
+38      37      -2.63   Constraint solver largest system.LargestNumTypeVariablesBound
+79      78      -1.27   Constraint solver largest system.LargestNumUnsimplifiedConstraints
+2593    2592    -0.04   Constraint solver overall.OverallNumStatesExplored
+589     588     -0.17   Constraint solver overall.OverallNumTypeVariableBindings
+482     481     -0.21   Constraint solver overall.OverallNumTypeVariablesBound
+972     971     -0.1    Constraint solver overall.OverallNumUnsimplifiedConstraints
+2593    2592    -0.04   ConstraintSystem.incrementScopeCounter
+2948    2947    -0.03   Generic signature builder.NumArchetypeAnchorCacheHits
+14767   15924   7.84    Generic signature builder.NumConformanceConstraints
+9701    10858   11.93   Generic signature builder.NumConformances
+5013    4241    -15.4   Generic signature builder.NumPotentialArchetypes
+5776    4084    -29.29  Generic signature builder.NumSameTypeConstraints
+...
+```
+
+#### Artifact-analysis tools
+
+Many performance issues manifest in the object files or module files produced by
+the Swift compiler -- say, by generating too much code -- so it can sometimes be
+helpful to look at the files the compiler outputs directly. The following tools
+are helpful in such cases:
+
+  - `llvm-objdump`, `llvm-otool` and `llvm-size` (which are LLVM-project
```

Comment:
nm is useful as well.

Reply:
And let's not forget about `utils/cmpcodesize` which is very useful when comparing two different versions of the same object/binary file. In particular, the `-c` and `-l` options are very interesting.

Reply:
Added mention of both tools.